src/main/scala/Validator.scala

```diff
@@ -147,76 +391,315 @@ class Validator(var schemaUri: Option[String], sourceUri: String = "") {
+      sqrt(parallelism).floor
+    )).floor.toInt
+
+    Source
+      .fromIterator(() => parser.asScala.iterator)
+      .filter(row => row.getRecordNumber > dialect.skipRows)
+      .grouped(rowGrouping)
+      .mapAsyncUnordered(degreeOfParallelism)(csvRows =>
+        Future {
+          csvRows.map(parseRow(schema, tableUri, dialect, table, _))
+        }
       )
-      .foreach {
-        case (row, validateRowOutput) => {
-          errors ++= validateRowOutput.warningsAndErrors.errors
-          warnings ++= validateRowOutput.warningsAndErrors.warnings
-          setChildTableForeignKeys(
-            validateRowOutput,
-            childTableForeignKeys
+      .fold[TableStateWithPrimaryKeyHashes](
+        ArrayBuffer.empty[WarningWithCsvContext],
+        ArrayBuffer.empty[ErrorWithCsvContext],
+        Map(),
+        Map(),
+        Map()
+      ) {
+
+        /** The last item in AccumulatedValues is used to store the hashes of every primary keys and its row numbers.
+          * To validate primary key efficiently, we need to have them in memory and HashSets gave the best performance.
+          * Storing all of the primary keys in sets leads to huge memory usage by the application.
+          * To be memory efficient, we hash the primary keys and then store them along with the rowNumbers.
+          * By hashing primary key values we're only identifying possible duplicates and it can be an overestimate of
+          * the actual number of duplicates because of hash collisions. Hash Collisions are also addressed at later stage
+          *
+          * {
+          *  56234234234: {1, 20},
+          *  45233453453: {2},
+          *  234234234234: {345}
+          * }
+          */
+
+        case (
+              (
+                warnings,
+                errors,
+                childTableForeignKeys,
+                parentTableForeignKeys,
+                mapPrimaryKeyHashToRowNumbers
+              ),
+              rowOutputs: Seq[ValidateRowOutput]
+            ) => {
+          extractInformationFromValidateRowOutputs(
+            warnings,
+            errors,
+            childTableForeignKeys,
+            parentTableForeignKeys,
+            mapPrimaryKeyHashToRowNumbers,
+            rowOutputs
           )
-          setParentTableForeignKeyReferences(
-            validateRowOutput,
-            parentTableForeignKeyReferences
+
+          (
+            warnings,
+            errors,
+            childTableForeignKeys,
+            parentTableForeignKeys,
+            mapPrimaryKeyHashToRowNumbers
           )
-          validatePrimaryKey(allPrimaryKeyValues, row, validateRowOutput)
-            .ifDefined(e => errors :+= e)
         }
       }
+      .flatMapConcat(keyCheckAggregateValues => {
+        // function getParser will return a parser here. If not, if wouldn't have reached here
+        val newParser = getParser(tableUri, dialect, format)
+          .getOrElse(throw new IllegalArgumentException)
```

Comment:
Can you pass a string into the `IllegalArgumentException` to specify what's wrong?

Reply:
Done